### DISCUSSION TITLE
Issue with EF Global Query Filters

### DISCUSSION THREAD

**Discussion by ShawInnes** [thepirat000/Audit.NET#629] [Q&A] [answered]
I recently had an issue which I _think_ is related to [Global Query Filters](https://learn.microsoft.com/en-us/ef/core/querying/filters) on EF.   I won't post it as issue until I can provide more useful info.

What I think happened:
- I implemented Soft Deletes using Query Filters.  Similar to the example in the above documentation.
```
modelBuilder.Entity<User>().HasQueryFilter(p => !p.IsDeleted);
```
- I have Audit.NET configured on my EntityFramework Context using `AuditSaveChangesInterceptor` and writing entries to the same Context using the `UseEntityFramework` extension method.
- The above has been working fine, logging audit records.
- A user record was Soft Deleted
- Subsequently an attempt was made to add that user again, but due to a small bug in my code, it caused a Unique Index violation on the underlying database (Postgres).  
- This threw a `Microsoft.EntityFrameworkCore.DbUpdateException`
- This exception seemed to cause a cascading failure within either Audit.NET, the AuditSaveChangesInterceptor, or EntityFramework Core. Each time the exception was thrown it would seem to retry, and then add another AuditEntity entry to the change set, fail again (because of the DbUpdateException on my User), and then add another AuditEntity entry.  
- Repeat until 💥 

The above happens instantly upon calling the .SaveChangesAsync().  If I remove the Interceptor everything is okay, so I know it's somewhat related to Audit.NET.

I can't see any configuration relating to retries or exception handling in the Audit.NET configuration, but I may be looking in the wrong place.

Any tips on where to continue/start looking would be great.

**Comment by thepirat000** [comment, marked as answer]
Maybe the same issue described here https://github.com/thepirat000/Audit.NET/issues/168

Can you share your configuration code (`UseEntityFramework()`)?

**Reply by ShawInnes** [reply]
Thanks @thepirat000 - such a quick response.  I was just searching for the wrong keywords, so thank you for pointing me in the right direction!


For reference this was my UseEntityFramework():

```
        Audit.Core.Configuration.Setup()
            .UseEntityFramework(ef => ef
                .AuditTypeMapper(t => typeof(AuditLog))
                .AuditEntityAction<AuditLog>((ev, entry, entity) =>
                {
                    entity.AuditData = entry;
                    entity.EntityType = entry.EntityType.Name;
                    entity.AuditTime = Instant.FromDateTimeUtc(ev.EndDate!.Value.ToUniversalTime());
                    entity.AuditUser = ev.Environment.UserName;
                    entity.AuditAction = entry.Action;
                    entity.TablePk = entry.PrimaryKey.First().Value.ToString()!;
                })
                .IgnoreMatchedProperties()
            );
```

I resolved my issue by including this code.  Adding the UseDbContext didn't work for me because my EF context doesn't have a parameterless constructor and I'm injecting it everywhere using DI.  I could probably dig further into that as a proper solution, but for now the below worked perfectly.

```
        Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaving, scope =>
        {
            if (scope.Event.Environment.Exception != null) 
                scope.Discard();
        });

```
